cmd/ingester/app/consumer/consumer.go

```diff
@@ -42,55 +42,55 @@ type Consumer struct {
 	internalConsumer consumer.Consumer
 	processorFactory ProcessorFactory
 
-	close    chan struct{}
-	isClosed sync.WaitGroup
+	partitionIDToState map[int32]*consumerState
+}
+
+type consumerState struct {
+	state             sync.WaitGroup
+	partitionConsumer sc.PartitionConsumer
 }
 
 // New is a constructor for a Consumer
 func New(params Params) (*Consumer, error) {
 	return &Consumer{
-		metricsFactory:   params.Factory,
-		logger:           params.Logger,
-		close:            make(chan struct{}, 1),
-		isClosed:         sync.WaitGroup{},
-		internalConsumer: params.InternalConsumer,
-		processorFactory: params.ProcessorFactory,
+		metricsFactory:     params.Factory,
+		logger:             params.Logger,
+		internalConsumer:   params.InternalConsumer,
+		processorFactory:   params.ProcessorFactory,
+		partitionIDToState: make(map[int32]*consumerState),
 	}, nil
 }
 
 // Start begins consuming messages in a go routine
 func (c *Consumer) Start() {
-	c.isClosed.Add(1)
-	c.logger.Info("Starting main loop")
-	go c.mainLoop()
+	go func() {
+		c.logger.Info("Starting main loop")
+		for pc := range c.internalConsumer.Partitions() {
+			if p, ok := c.partitionIDToState[pc.Partition()]; ok {
+				p.state.Wait()
+				delete(c.partitionIDToState, pc.Partition())
```

Comment:
technically not needed as the line below overwrites it, but being explicit is good too